### ISSUE TITLE
Problems with world gen when generating on multiple threads

### ISSUE THREAD

**Issue by nikolaus-albinger** [minetest/minetest#11986]
5.4.1, Irrlicht 1.8.4, LuaJIT 2.1.0-beta3

##### OS / Hardware
Operating system: Windows 10 / Linux Mint 20.3
CPU: i5-10400F
GPU model: Nvidia GeForce RTX 3060
OpenGL version: 4.6.0

##### Summary
I'm not sure if it's really a bug in the engine.
When I generate worlds with multiple threads (num_emerge_threads) I get strange errors. These include caves in which "slabs of stone" are generated and rarely half trees. In Mcl2, this effect is even more noticeable at the end.
![1](https://user-images.githubusercontent.com/89982526/150816737-743cf6fe-6ae3-4781-a6bf-3dbe82deab60.png)

**Comment by erlehmann** [comment]
> When I generate worlds with multiple threads (num_emerge_threads) I get strange errors. These include caves in which "slabs of stone" are generated and rarely half trees. In Mcl2, this effect is even more noticeable at the end.

We call them ”stone clouds” and you can also find them under the overworld bedrock in the void.
If you are really lucky, you can find an entire 80x80x80 mapblock made out of stone underground.
(Underground includes the End and the Nether, as they are located very deep below the overworld.)

Emerging stuff is pretty racy AFAIK. You may be able to figure out more using ThreadSanitizer.
I believe @kay27 has dealt with this stuff a lot and implemented a lot of workarounds for it.

In the meantime, try using a single thread. :wink:

**Comment by sfan5** [comment]
This is a duplicate of #9357 and/or #5618.

**Comment by kay27** [comment]
My advice would be to use one thread and try to make mapgen work faster on single thread. In [this PR](https://git.minetest.land/MineClone2/MineClone2/pulls/1976) you can find a mod `CORE/mcl_mapgen` with the API. The idea of the API is to deal the world only after grinder-like chunk-in-shell feature really freed the block (or the chunk) and really won't further modify.

Old-style APIs forced you to place same things up to 8 times per a block, or to emerge way larger areas - all was slow. So, new API allows you to make you mapgen work 2 times faster by using suitable calls from it.

I'm afraid, if you want to use engine multithreaded mapgen still, it will probably require changes in Lua code - removing persistent lvm buffers, etc. The code is not multithreaded right now. I suspect, on engine level, `chunk-in-shell` could block the multithreading too. So the new API hopefully will let us live with single threads for some time. In general, engine-produced landscapes look really good, I'm not sure if we need to fix anything there right now.

Further reading: https://git.minetest.land/MineClone2/MineClone2/issues/1395
